### ISSUE TITLE
[feature request] option to enforce one root

### ISSUE THREAD

**Issue by mvayngrib** [mafintosh/merkle-tree-stream#3]
e.g. in bitcoin, at any level of the tree where there is not an even number of nodes, the odd leaf will get hashed with itself so that there is finally just one merkle root

**Comment by mafintosh** [comment]
you can access the current roots using `streams.roots`. after `end` has been emitted you can hash these together to get a single root using what ever scheme you want

**Comment by mvayngrib** [comment]
right, i was just wondering if you wanted to have it as an option on end() or something. Or better on the constructor, to not break the end() native stream logic

**Comment by mafintosh** [comment]
Ah I see. What index would be last root have? The parent index of the left most "peak" ?

**Comment by mvayngrib** [comment]
that sounds right to me

**Comment by mafintosh** [comment]
okay. yea an optional function like finalize that took an array of roots and returned a hash would be ok with me. if you wanna send a pr for that i'll merge it :)

**Comment by mvayngrib** [comment]
i think the finalize function should actually be pushing nodes into the stream, not just returning a hash. Let me put a PR together and we can continue the discussion there

**Comment by mafintosh** [comment]
yep i agree. i meant that the function provided as an option should just return a hash and then the module would push a node to the stream

**Comment by mvayngrib** [comment]
i was thinking rather that roots would just end up with length 1. I think we have different pictures in our heads :)

**Comment by mvayngrib** [comment]
created [merkle-proofs](https://github.com/mvayngrib/merkle-proofs) based on the finalize branch. So far so good.
